File: c++/Graphs/Swim in Rising Water.cpp
// // Swim in Rising Water
// Problem Statement : You are given an n × n matrix grid where grid[i][j] is the unique elevation of cell(i, j).Rain starts falling at time t = 0. At any later time t ≥ 0, every cell is covered by water to depth t.You may move 4 - directionally(up, down, left, right) between adjacent cells instantaneously iff the elevations of both cells are ≤ t.Starting from the top - left cell(0, 0), return the minimum time t at which you can reach the bottom - right cell(n − 1, n − 1).

#include <bits/stdc++.h>
using namespace std;

class Solution
{
public:
    // Minimum time to reach bottom right cell
    int swimInWater(vector<vector<int>> &grid)
    {
        // Get grid size
        int n = grid.size();

        // Create a min-heap for cells based on elevation
        priority_queue<vector<int>, vector<vector<int>>, greater<vector<int>>> minHeap;

        // Create visited matrix
        vector<vector<int>> visited(n, vector<int>(n, 0));

        // Push starting cell to heap
        minHeap.push({grid[0][0], 0, 0});
        visited[0][0] = 1;

        // Direction vectors for movement
        vector<pair<int, int>> dirs = {{0, 1}, {1, 0}, {0, -1}, {-1, 0}};

        // Process heap until destination is reached
        while (!minHeap.empty())
        {
            // Extract cell with least elevation
            auto curr = minHeap.top();
            minHeap.pop();
            int elevation = curr[0], r = curr[1], c = curr[2];

            // If destination reached, return elevation
            if (r == n - 1 && c == n - 1)
                return elevation;

            // Explore neighboring cells
            for (auto &dir : dirs)
            {
                int nr = r + dir.first;
                int nc = c + dir.second;

                // Check bounds and if not visited
                if (nr >= 0 && nc >= 0 && nr < n && nc < n && !visited[nr][nc])
                {
                    visited[nr][nc] = 1;
                    minHeap.push({max(elevation, grid[nr][nc]), nr, nc});
                }
            }
        }
        return -1;
    }
};

int main()
{
    vector<vector<int>> grid = {
        {0, 2}, {1, 3}};
    Solution sol;
    cout << "Minimum time to reach destination: " << sol.swimInWater(grid) << endl;
    return 0;
}

File: java/778-swim-in-rising-water/swim-in-rising-water.java
class Solution {
    boolean[][] visited;
    public int swimInWater(int[][] grid) {
        int n = grid.length;
        int l = 0; int r = n * n - 1;
        int min = n * n - 1;

        while(l <= r) {
            int mid = l + ((r - l)/2);
            visited = new boolean[n][n];
            if(dfs(grid, mid, 0, 0, n)){
                r = mid - 1;
                min = mid;
            }
            else
                l = mid + 1;
        }
        return min;
    }
    private boolean dfs(int[][] grid, int max_value, int i, int j, int n){
        if(i < 0 || j < 0 || j == n || i == n || visited[i][j] || grid[i][j] > max_value) //break
            return false;
        if(i == n-1 && j == n-1)
            return true;

        visited[i][j] = true;   // mark visited

        boolean right = dfs(grid, max_value, i, j+1, n);
        boolean left = dfs(grid, max_value, i, j-1, n);
        boolean bottom = dfs(grid, max_value, i+1, j, n);
        boolean top = dfs(grid, max_value, i-1, j, n);

        return (right || left || bottom || top) ? true : false;
    }
    
}
